src/sqlfluff/rules/L003.py

```diff
@@ -244,28 +349,30 @@ def _eval(self, segment, raw_stack, memory, **kwargs):
                 "comment_lines": [],
             }
 
-        if segment.is_type("newline"):
-            memory["in_indent"] = True
-            # We're not going to flag on empty lines so we can safely proceed
-            return LintResult(memory=memory)
-        elif memory["in_indent"]:
-            if segment.is_type("whitespace"):
-                # it's whitespace, carry on
-                return LintResult(memory=memory)
-            elif segment.segments or (segment.is_meta and segment.indent_val != 0):
-                # it's not a raw segment or placeholder. Carry on.
-                return LintResult(memory=memory)
-            else:
-                memory["in_indent"] = False
-                # we're found a non-whitespace element. This is our trigger,
-                # which we'll handle after this if-statement
-        else:
-            # Not in indent and not a newline, don't trigger here.
+        # Is this the last segment? If so, need to "flush" any leftovers.
+        is_last = self._is_last_segment(segment, memory, parent_stack, siblings_post)
+
+        if not segment.is_type("newline") and not is_last:
+            # We only process complete lines or on the very last segment
+            # (since there may not be a newline on the very last line)..
             return LintResult(memory=memory)
 
         res = self._process_raw_stack(
-            raw_stack + (segment,), tab_space_size=self.tab_space_size
+            raw_stack, tab_space_size=self.tab_space_size, templated_file=templated_file
         )
+        if res:
+            return self._process_current_line(res, memory, templated_file)
+        else:
+            return LintResult(memory=memory)
+
+    def _process_current_line(self, res, memory, templated_file):
```

Comment:
I split this function out of the original `_eval()` function to make the code a little easier to understand and to separate concerns a bit. The main `_eval()` function processes each segment as it's passed in. OTOH, this function is only executed at the end of each **line** of code, to check the indentation.